File: leetcode/go/900.rle-iterator.go
type RLEIterator struct {
    encoding []int
    i int
}


func Constructor(encoding []int) RLEIterator {
    return RLEIterator {
        encoding: encoding,
        i: 0,
    }
}


func (this *RLEIterator) Next(n int) int {
    if this.i >= len(this.encoding) {
        return -1;
    }
    ans, count := this.encoding[this.i+1], 0
    for n > 0 {
        for this.encoding[this.i] == 0 {
            this.i += 2
            if this.i >= len(this.encoding) {
                return -1
            }
            ans = this.encoding[this.i+1]
        }
        count = min(n, this.encoding[this.i])
        this.encoding[this.i] -= count
        n -= count
    }
    return ans
}

func min(a, b int) int {
    if a < b {
        return a
    }
    return b
}

/**
 * Your RLEIterator object will be instantiated and called as such:
 * obj := Constructor(encoding);
 * param_1 := obj.Next(n);
 */

File: leetcode/rust-leetcode/src/900.rle-iterator.rs
struct RLEIterator {
    encoding: Vec<i32>,
    i: usize,
}


/** 
 * `&self` means the method takes an immutable reference.
 * If you need a mutable reference, change it to `&mut self` instead.
 */
impl RLEIterator {

    fn new(encoding: Vec<i32>) -> Self {
        Self {encoding, i: 0}
    }
    
    fn next(&mut self, mut n: i32) -> i32 {
        if self.i >= self.encoding.len() {
            return -1;
        }
        let mut ans = self.encoding[self.i+1];
        let mut count = 0;
        while n > 0 {
            while self.encoding[self.i] == 0 {
                self.i += 2;
                if self.i >= self.encoding.len() {
                    return -1;
                }
                ans = self.encoding[self.i + 1];
            }
            count = n.min(self.encoding[self.i]);
            self.encoding[self.i] -= count;
            n -= count;
        }
        ans
    }
}

/**
 * Your RLEIterator object will be instantiated and called as such:
 * let obj = RLEIterator::new(encoding);
 * let ret_1: i32 = obj.next(n);
 */

